Implement parser

kroppt opened the pull request:
The parser runs concurrently. The input, output, and error channels are passed into the constructor.

TODO
- [ ] Add more negative tests
- [x] Add a way to append the input
- [x] Add error handling / signalling

kroppt committed:
Add basic state diagram and grammar document

kroppt committed:
Add parser message type, begin and tag states

kroppt committed:
Rename parsing helper function

kroppt committed:
Fix parseUntil rewind logic

kroppt committed:
Speed up parseUntil

kroppt committed:
Move things around and aesthetic changes

kroppt committed:
Add prefix state

kroppt committed:
Add command state

kroppt committed:
Set command in commandState

kroppt committed:
Add param state

kroppt committed:
Add trailing param state

kroppt committed:
Add end state

kroppt committed:
Add basic command tests and cleanup

kroppt committed:
Add first param test and fix param parsing

kroppt committed:
Refactor parser tests to table and match expected

kroppt committed:
Add timeout for parser tests

kroppt committed:
Check for closed channel in parser tests

kroppt committed:
Add param examples 2, 3, 4, and fix prefix parsing

kroppt committed:
Add param example 5 and fix param parsing

kroppt committed:
Add complete examples 1, 2, and fix tag parsing

kroppt committed:
Add complete example 3

kroppt committed:
Add tag example 1 and fix tag parsing

kroppt committed:
Add tag example 2

kroppt committed:
Add tag examples 3, 4, and fix tag vendor parsing

kroppt committed:
Add complete example 4

kroppt committed:
Add parsing failure test helper

kroppt committed:
Add command parse fail tests, verify command input

kroppt committed:
Add tag key length verification

kroppt committed:
Add a few more parse fail tests

kroppt committed:
Add parsing over multiple spaces and tests for it

kroppt committed:
Improved failure case in positive parsing testing

Now, we can check for failure without having to time out.

kropptrevor commented:
Can you share the state machine diagram?
Also, what syntax is used for grammar.txt?

kroppt committed:
Change main loop and add input receiving

kroppt committed:
Preserve state for clean pausing and resuming

kroppt commented:
@kropptrevor State diagram is a part of the PR. The grammar uses augmented Backus-Naur form (ABNF), which I got from https://modern.ircdocs.horse/.

kroppt committed:
Implement concurrent error handling

When encountering an error, the following happens:
- sends message over error channel with partially constructed message
- throws out everything in input up to \r\n
- resets to begin state

When the parser is in begin state and runs out of input, it pauses.
When new input is fed into it, it is automatically resumed.
This feature is internal and different from cancellation.

kropptrevor commented on parser/parser.go:
```diff
@@ -0,0 +1,599 @@
+		if r == '!' {
+			p.Consume()
+			p.msg.Prefix.Username = parseUntil(p, func(r rune) bool {
+				return r != '@'
```
Consider `thatserver/johnny!johnnyman sometext@thathost`.
The username would be parsed as `johnnyman sometext`.
Are spaces allowed for usernames?

kroppt replied:
Done. There is no concrete specification, so I made one up based on what I could test on some random IRC server.

kroppt committed:
Add specification for usernames and implement it

kropptrevor commented on parser/parser.go:
```diff
@@ -0,0 +1,629 @@
+
+// Error is
+type Error struct {
+	Partial Message
```
What is the purpose of `Partial Message`?

kroppt replied:
It's mainly just context for the state error. There are multiple ways to get to certain states, so having this info helps figure that out.

kropptrevor commented on parser/parser.go:
```diff
@@ -0,0 +1,629 @@
+			lastRuneCR = true
+			return true
+		case unicode.IsLetter(r):
+			fallthrough
```
Why does the case for when it is a letter not set `lastRuneCR`?

kroppt replied:
I fallthrough to the case where it's set to false, i.e. the `isNumber` case. The reason it's false is because the CR case is checked as an earlier case, so we know the rune cannot be CR in any other case, such as this one.

kroppt committed:
Move tests into separate package
